Add syntax for generative functor application

goldfirere opened the pull request:
Previously, writing `F ()` was the same as writing `F (struct end)`, even though the latter looks like the use of an applicative functor, not a generative one.

This commit, originally written by Frédéric Bour <fred@tarides.com>, adds new syntax to our AST to represent generative functor application and propagates this change throughout the compiler.

In addition, it adds a new warning, 73, to report when a user has written `F (struct end)` but should now update to `F ()`.

let-def committed:
Add syntax for generative functor application

Previously, writing [F ()] was the same as writing
[F (struct end)], even though the latter looks like the
use of an applicative functor, not a generative one.

This commit, originally written by Frédéric Bour <fred@tarides.com>,
adds new syntax to our AST to represent generative functor
application and propagates this change throughout the compiler.

In addition, it adds a new warning, 73, to report when a user
has written [F (struct end)] but should now update to [F ()].

goldfirere commented:
This is the continuation of #11456.

dra27 commented on utils/warnings.ml:
```diff
@@ -527,6 +529,11 @@ let descriptions = [
+    names = ["generative-application-expects-unit"];
+    description = "A generative functor is applied to an empty structure \
+                   (struct end) rather than to ().";
+    since = since 4 15 };
```
```suggestion
    since = since 5 1 };
```

smuenzel commented on utils/warnings.ml:
```diff
@@ -1122,12 +1129,15 @@ let message = function
   | Tmc_breaks_tailcall ->
       "This call\n\
-       is in tail-modulo-cons positionin a TMC function,\n\
+       is in tail-modulo-cons position in a TMC function,\n\
```
Should we move this correction to a separate pr, so the diff doesn't have unrelated content?

goldfirere replied:
For something as small and obviously correct as this, I think the overhead of doing this work is greater than the value it delivers. I agree in general about unrelated work, but there is a real overhead to separating it out.

goldfirere committed:
Update utils/warnings.ml

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

goldfirere committed:
Add a Change

Octachron commented on typing/typemod.ml:
```diff
@@ -2260,59 +2267,83 @@ and type_module_aux ~alias sttn funct_body anchor env smod =
+        let args = List.map simplify_app_summary args in
+        let mty_f = md_f.mod_type in
+        let lid_app = None in
+        raise(Includemod.Apply_error {loc=apply_loc;env;lid_app;mty_f;args})
```
I would be in favor of adding an `apply_error ()` function to factorize the construction of the `Apply_error` in the unit application and standard one.

Octachron commented on typing/typemod.ml:
```diff
@@ -2260,59 +2267,83 @@ and type_module_aux ~alias sttn funct_body anchor env smod =
+        | Some arg ->
+          if arg.is_syntactic_unit then
+            Location.prerr_warning arg.arg.mod_loc
+              Warnings.Generative_application_expects_unit
```
We could update the warning scope (with `Builtin_attributes.warning_scope`) here to ensure that
```ocaml
module M = F(struct end[@warning "-73"])
```
disables the warning. 

I was wondering if it might help `ppxlib` (but maybe ppxlib would have no choice other than disabling the globally warning whenever translating to or from a previous AST).

panglesd replied:
If I am not missing anything, `ppxlib` might need to disable the warning on [`Ast_builder.pmod_apply`](https://ocaml.org/p/ppxlib/0.28.0/doc/Ppxlib/Ast_builder/Default/index.html#val-pmod_apply)-generated nodes, but won't need to disable it globally.

Ensuring that adding the attribute `[@warning "-73"]` on the functor argument disables the warning would be great if we decide to disable the warning on [`Ast_builder.pmod_apply`](https://ocaml.org/p/ppxlib/0.28.0/doc/Ppxlib/Ast_builder/Default/index.html#val-pmod_apply)-generated nodes.

Octachron commented:
cc @pitag-ha @panglesd : I am not completely sure how this change will affect `ppxlib`: this is a parsetree changes that splits an existing node `F(...)` in two (`F(A)` and `F()`), accompanied by a new warning when the old version of the node is used instead of the new specific node (`F(struct end)` vs `F()`). As far as I understand, this means that ppxlib would have to essentially disable the warning for as long as its reference parsetree is not updated to the new parsetree. Is that the case?

Also would it help the ppxlib team if we had an `parsetree-change` label?

panglesd commented:
Thanks @Octachron for the ping!

About this specific issue:

I don't think the problem appears when there are a parsetree migrations: old version -> 5.1 -> old version. There might be `F(struct end)` warnings in the 5.1 parsetree used for rewriting, but those would disappear when migrating back to the old version.
However, you are right that any PPX that was generating some generative functor (by creating a `F(struct end)` node) would generate a warning when used on 5.1. Disabling the warning on [`Ast_builder.pmod_apply`](https://ocaml.org/p/ppxlib/0.28.0/doc/Ppxlib/Ast_builder/Default/index.html#val-pmod_apply)-generated nodes is a good solution to ensure compatiblity.

On the other hand, according to [sherlocode](https://sherlocode.com/?q=pmod_apply), this function is applied only a few times, and from what I understand, never with an empty `struct`. So, the change in this PR won't break code that is referenced by sherlocode.
That means that not adding a warning is also fine, since it prevents future users to incorrectly use `pmod_apply` with an empty struct instead of `pmod_apply_unit`, without breaking current users.

About the `parsetree-change` label:

Thanks, that's a very good idea! I think that would be great both for PPX authors and the ppxlib team to be able to find all discussions about changed syntax (I often had to go to the history of the `parsetree.mli` for that). (Unfortunately it seems there is no way to get github notifications for a specific label, so a ping if needed would be also very appreciated.)

goldfirere committed:
address comments; F (struct end [@warning "-73"])

goldfirere commented:
Thanks for the review @Octachron!

I've addressed the comments above and was able to get the `F (struct end [@warning "-73"])` not to warn, as requested. (I agree that this is a good idea!)

Is there anything else outstanding here?

goldfirere committed:
Actually remember to update test case

Octachron commented on typing/typemod.ml:
```diff
@@ -1814,7 +1814,7 @@ exception Not_a_path
   match mexp.mod_desc with
   | Tmod_ident (p,_) -> p
-  | Tmod_apply(funct, arg, _coercion) when !Clflags.applicative_functors ->
+  | Tmod_apply(funct, arg, _coe) when !Clflags.applicative_functors ->
```
Nitpick: I would rather keep the history since this code does not need to be updated for `Tmod_apply_unit`.
Optionally, it might improve readability to explode the wildcard below.

Octachron commented on typing/typemod.ml:
```diff
@@ -2260,59 +2267,86 @@ and type_module_aux ~alias sttn funct_body anchor env smod =
+      begin match app_view with
+      | { arg = None; _ } -> apply_error ()
+      | { loc = app_loc; attributes = app_attributes;
+          arg = Some { shape = app_shape; path = arg_path; arg } } ->
```
`app_shape` ⇒ `arg_shape` ?
(since it is the shape of the argument and not of the application which is computed later)

Octachron reviewed (approved):
Overall, I like that this PR explicit the difference between `F()` and `F(struct end)` which reflects the already distinct rules for the definition of applicative and generative functors.

In particular, this PR removes one of the case where it was not possible to lift a constant to separate definition:
```ocaml
module Ok = F(struct end)
module Constant = struct end
module Wrong = F(Constant)
```

The improvement to the lambda IR was a bit unexpected, but it is nice and straightforward.

The PR looks ready to merge once the changelog is updated, even if I have two forgettable remarks.

goldfirere committed:
Final comments from @Octachron

Octachron commented on typing/typemod.ml:
```diff
@@ -1814,11 +1814,13 @@ exception Not_a_path
   match mexp.mod_desc with
   | Tmod_ident (p,_) -> p
-  | Tmod_apply(funct, arg, _coercion) when !Clflags.applicative_functors ->
+  | Tmod_apply(funct, arg, _coercon) when !Clflags.applicative_functors ->
```
_coercon ⇒ _coerc**i**on

goldfirere replied:
D'oh! Thanks!

goldfirere committed:
Spell correction

gasche commented:
Merged, thanks!
